scripts/validate_pinned_versions.py

```
@@ -0,0 +1,71 @@
+        # drop extras for PyPI project endpoint
+        package = package_part.split("[", 1)[0].strip()
+        version = version.strip()
+        if package and version:
+            yield line_number, package, version
+
+
+def _pypi_versions(package: str) -> set[str]:
+    """Return all published versions for a package from the PyPI JSON API."""
+    url = f"https://pypi.org/pypi/{package}/json"
+    request = urllib.request.Request(url, headers={"User-Agent": "micro-mart-pin-validator"})
+    with urllib.request.urlopen(request, timeout=15) as response:
```

Ruff (0.15.12)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)